### PR TITLE
ARROW-10628: [Rust] flag clippy warnings as errors

### PR DESCRIPTION
**Pull request by alamb**
# Rationale

We do currently run clippy (aka rust linter) on all PRs. However, the way we invoke clippy means that any clippy warnings introduced in PRs do not cause a CI failure

I believe the intended behavior is that we keep the arrow project "Clippy clean"

# Changes
* Upgrade the rust_lint.sh script to test all targets and treat warnings as fatal
* Updates the code so that the project passes all clippy lints 🎉 

# Example of this not working today
For example, this PR has clippy flag  a warning and yet the test still passes (note the PR doesn't introduce the warning).

https://github.com/apache/arrow/runs/1408182529?check_suite_focus=true

```
    Checking arrow v3.0.0-SNAPSHOT (/Users/runner/work/arrow/arrow/rust/arrow)
warning: unneeded unit return type
   --> arrow/src/buffer.rs:694:45
    |
694 |     pub fn resize(&mut self, new_len: usize) -> () {
    |                                             ^^^^^^ help: remove the `-> ()`
    |
    = note: `#[warn(clippy::unused_unit)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unused_unit

warning: 1 warning emitted

    Finished dev [unoptimized + debuginfo] target(s) in 1m 11s
~/work/arrow/arrow
```

# Testing of this PR:
*  The initial commit of this PR f50ca5a only includes the updated clippy command and it fails CI (see https://github.com/apache/arrow/pull/8687/checks?check_run_id=1408834360)
* When the second commit to fix the clippy warning is introduced, then CI passes: https://github.com/apache/arrow/pull/8687/checks?check_run_id=1409016544

### CODE REVIEW

File: ci/scripts/rust_lint.sh [hunk 1]
```diff
@@ -17,10 +17,14 @@
...
+#
+# Usage: rust_lint.sh <arrow checkout directory>
+
 set -e
 
 source_dir=${1}/rust
 
 pushd ${source_dir}/arrow
-    cargo clippy -- -A clippy::redundant_field_names
+    cargo clippy --all-targets --workspace -- -D warnings -A clippy::redundant_field_names
```
Comment (alamb) [inline ci/scripts/rust_lint.sh:29]: The `-D warnings` means "disallow anything that clippy would warn against" and treat them as an error

File: rust/parquet_derive/src/parquet_field.rs [hunk 2]
```diff
@@ -252,6 +252,8 @@ impl Field {
     }
 }
 
+#[allow(clippy::enum_variant_names)]
```
Comment (alamb) [inline rust/parquet_derive/src/parquet_field.rs:255]: While clippy has some valid points with these warnings, I did not want to address them in this PR

**Review by jorgecarleitao — APPROVED** [review]
LGTM!

**Comment by jorgecarleitao** [comment]
@alamb There was no issue 10636 on Jira, so I created one, assigned it to you, and assigned this PR to it.

**Comment by alamb** [comment]
Thank you @jorgecarleitao !

**Comment by alamb** [comment]
FYI @nevi-me

**Pull request closed without merging**